source/extensions/filters/http/ext_proc/processor_state.h

```diff
@@ -16,29 +20,50 @@ namespace ExternalProcessing {
 
 class Filter;
 
+class QueuedChunk {
+public:
+  // True if this represents the last chunk in the stream
+  bool end_stream = false;
+  // True if the chunk was actually sent to the gRPC stream
+  bool delivered = false;
+  Buffer::OwnedImpl data;
+};
+using QueuedChunkPtr = std::unique_ptr<QueuedChunk>;
+
 class ProcessorState : public Logger::Loggable<Logger::Id::filter> {
 public:
-  // This describes whether the filter is waiting for a response to a gRPC message
+  // This describes whether the filter is waiting for a response to a gRPC message.
+  // We use it to determine how to respond to stream messages send back from
+  // the external processor.
   enum class CallbackState {
     // Not waiting for anything
     Idle,
-    // Waiting for a "headers" response
+    // Waiting for a "headers" response. This may be true even if we started
+    // to receive data frames for a message.
     HeadersCallback,
-    // Waiting for a "body" response in buffered mode
+    // Waiting for a "body" response in buffered mode.
     BufferedBodyCallback,
-    // and waiting for a "trailers" response
+    // Waiting for a "body" response in streaming mode.
+    StreamedBodyCallback,
+    // Waiting for a "body" response in streaming mode in the special case
+    // in which the processing mode was changed while there were outstanding
```

Comment:
When would we sensible change the processing mode during streaming?

Reply:
Sensibly, I could see a case in which a filter might want to inspect only the beginning of a large message body. As another example, I was thinking of a caching solution recently in which we'd want to cache message bodies, but only if the message body is smaller than a certain size. Once a processor sees that the body will be too large, it might want to change the processing mode to streamline things if it no longer cares.

The other reason is because we allow that in the protocol now. If we decide instead to ignore processing mode changes while streaming a body then we likely need to add some more logic elsewhere.

Reply:
@gbrail it could be especially useful if you're looking for a specific substring in a request / response body, and once you found it you can ignore the rest (since a policy has been already activated)